code/modules/plumbing/plumbers/reaction_chamber.dm

```diff
@@ -131,8 +105,79 @@
 				. = TRUE
 			if(.)
 				target_temperature = clamp(target, 0, 1000)
+
+//Chemistry version of reaction chamber that allows for acid and base usage
+/obj/machinery/plumbing/reaction_chamber/chem
+	name = "reaction chamber"
+
+	///If above this pH, we start dumping buffer into it
+	var/acidic_limit = 9
+	///If below this pH, we start dumping buffer into it
+	var/alkaline_limit = 5
+
+	///Beaker that holds the acidic buffer. I don't want to deal with snowflaking so it's just a separate thing. It's a small (50u) beaker
+	var/obj/item/reagent_containers/cup/beaker/acidic_beaker
+	///beaker that holds the alkaline buffer.
+	var/obj/item/reagent_containers/cup/beaker/alkaline_beaker
+
+/obj/machinery/plumbing/reaction_chamber/chem/Initialize(mapload, bolt, layer)
+	. = ..()
+
+	acidic_beaker = new (src)
+	alkaline_beaker = new (src)
+
+	AddComponent(/datum/component/plumbing/acidic_input, bolt, custom_receiver = acidic_beaker)
+	AddComponent(/datum/component/plumbing/alkaline_input, bolt, custom_receiver = alkaline_beaker)
+
+/// Make sure beakers are deleted when being deconstructed
+/obj/machinery/plumbing/reaction_chamber/chem/Destroy()
+	QDEL_NULL(acidic_beaker)
+	QDEL_NULL(alkaline_beaker)
+	. = ..()
+
+/obj/machinery/plumbing/reaction_chamber/chem/process(delta_time)
+	//add acidic/alkaine buffer if over/under limit
+	if(reagents.is_reacting && reagents.ph < alkaline_limit)
+		alkaline_beaker.reagents.trans_to(reagents, 1 * delta_time)
+	if(reagents.is_reacting && reagents.ph > acidic_limit)
+		acidic_beaker.reagents.trans_to(reagents, 1 * delta_time)
+	. = ..()
+
+/obj/machinery/plumbing/reaction_chamber/chem/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "ChemReactionChamber", name)
+		ui.open()
+
+/obj/machinery/plumbing/reaction_chamber/chem/ui_data(mob/user)
+	var/list/data = list()
+
+	var/list/reagents_data = list()
+	for(var/datum/reagent/required_reagent as anything in required_reagents) //make a list where the key is text, because that looks alot better in the ui than a typepath
+		var/list/reagent_data = list()
+		reagent_data["name"] = initial(required_reagent.name)
+		reagent_data["required_reagent"] = required_reagents[required_reagent]
+		reagents_data += list(reagent_data)
+
+	data["reagents"] = reagents_data
+	data["emptying"] = emptying
+	data["temperature"] = round(reagents.chem_temp, 0.1)
+	data["ph"] = round(reagents.ph, 0.01)
+	data["targetTemp"] = target_temperature
+	data["isReacting"] = reagents.is_reacting
+	data["reagentAcidic"] = acidic_limit
+	data["reagentAlkaline"] = alkaline_limit
+	return data
+
+/obj/machinery/plumbing/reaction_chamber/chem/ui_act(action, params)
+	if(.)
+		return
+	. = TRUE
```

Comment:
This will never pass. '.' is the variable representing the return value if none is specified, which is FALSE by default. 
You then set it to TRUE and don't do anything with it? I can't really blame you because the parent proc is smoking crack aswell

Reply:
Tried to make this less insane by going off of other uses of ui_act, hopefully smokes less crack now

Reply:
Looks a lot better! Nice work